src/main/java/org/openstreetmap/atlas/checks/validation/linear/lines/WaterWayCheck.java

```diff
@@ -0,0 +1,526 @@
+     */
+    public boolean doesLineEndInOcean(final LineItem line)
+    {
+        final Atlas atlas = line.getAtlas();
+        final PolyLine linePolyline = line.asPolyLine();
+        final Location last = linePolyline.last();
+        if (atlas.areasCovering(last, this.validOceanTags::test).iterator().hasNext()
+                || atlas.areasCovering(last, this.oceanBoundaryTags::test).iterator().hasNext())
+        {
+            return true;
+        }
+        final List<LineItem> lines = new ArrayList<>();
+        atlas.lineItemsIntersecting(line.asPolyLine().bounds(), this.oceanBoundaryTags::test)
+                .forEach(lines::add);
+        final LineItem[] intersecting = lines.stream()
+                .filter(l -> l.asPolyLine().intersects(linePolyline)).toArray(LineItem[]::new);
+        final SegmentIndexComparator segmentComparator = new SegmentIndexComparator(linePolyline);
+        for (final LineItem lineItem : intersecting)
+        {
+            // Get way segments from the two lines that intersect
+            final Collection<Pair<Segment, Segment>> segs = getIntersectingSegments(linePolyline,
+                    lineItem.asPolyLine());
+            // Get the *latest* crossing segment of the waterway
+            // (just in case the waterway crosses a coastline multiple times)
+            final Segment max = segs.stream().map(Pair::getKey).distinct().max(segmentComparator)
+                    .orElse(null);
+            // Get the crossing segments of the coastline (probably just one)
+            final Collection<Segment> crosses = segs.stream()
+                    .filter(pair -> pair.getLeft().equals(max)).map(Pair::getValue)
+                    .collect(Collectors.toSet());
+            // Create a shortened coastline to use to check if the waterway ends inside the ocean
+            final PolyLine coast = new PolyLine(
+                    crosses.stream().flatMap(pCoast -> Stream.of(pCoast.first(), pCoast.last()))
+                            .toArray(Location[]::new));
+            // If the waterway ends to the right of the coastline, it ended in an ocean.
+            // If the waterway ends on the coastline, it ended in an ocean.
+            if (isRightOf(coast, linePolyline.last())
+                    || lineItem.asPolyLine().contains(linePolyline.last()))
+            {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    /**
+     * Check if a line ends in a waterway sink
+     *
+     * @param lineItem
+     *            The LineItem to check
+     * @return {@code true} if the waterway can reasonably be expected to end
+     */
+    public boolean doesLineEndInSink(final LineItem lineItem)
+    {
+        // If the way is a sink, it obviously ends in a sink...
+        if (this.waterwaySinkTagFilter.test(lineItem))
+        {
+            return true;
+        }
+        if (lineItem instanceof Line)
```

Comment:
I think we could have the same treatment of Lines and Edges here. If either feature has a node or point at its end that's a sink, return true

Reply:
This goes back to trying to optimize to avoid significant increases in processing time. Again, compared to my latest in-progress check, this is peanuts. :)